### PR TITLE
Backport "Fix healAmbiguous to compareAlternatives with disambiguate = true" to 3.5.0

### PR DESCRIPTION
**Pull request by WojciechMazur**
Backports #21339 to Scala 3.5.0-RC7

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 1]
```diff
@@ -1816,17 +1836,15 @@ trait Applications extends Compatibility {
             isAsGood(alt1, tp1.instantiate(tparams.map(_.typeRef)), alt2, tp2)
           }
         case _ => // (3)
-          def isGiven(alt: TermRef) =
-            alt1.symbol.is(Given) && alt.symbol != defn.NotGivenClass
-          def compareValues(tp1: Type, tp2: Type)(using Context) =
-            isAsGoodValueType(tp1, tp2, isGiven(alt1), isGiven(alt2))
+          def compareValues(tp2: Type)(using Context) =
+            isAsGoodValueType(tp1, tp2, alt1.symbol.is(Implicit))
```
Comment (WojciechMazur) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala:1840]: This change comes from https://github.com/scala/scala3/pull/21226 that was not backported to 3.5.0
@odersky Can you confirm that it's expected and should remain in 3.5?

File: tests/pos/i13044.scala [hunk 2]
```diff
@@ -1,4 +1,4 @@
-//> using options -Xmax-inlines:33
+//> using options -Xmax-inlines:35
```
Comment (WojciechMazur) [inline tests/pos/i13044.scala:1]: Change required when backported to 3.5.0 but not needed in main

File: tests/neg/scala-uri.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
+//> using options -source:3.6
```
Comment (WojciechMazur) [inline tests/neg/scala-uri.scala:1]: Backported version required -source:3.6 which is not needed in main

Comment (EugeneFlesselle) [reply tests/neg/scala-uri.scala:1]: main has the baseVersion set to 3.6 already, so we get the `CompareScheme.New` by default. 
While in 3.5 we maintain the intermediate comparison scheme, in which the test case used to be un-ambiguous, so I think this is as expected.

File: tests/pos/i13044.scala (hunk 2 above)
Comment (EugeneFlesselle) [reply tests/pos/i13044.scala:1]: Yes the same change was needed in 2facda005615361a701616adccd2d77fb5bea1d1

**Review by EugeneFlesselle — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 1 above)
Comment (WojciechMazur) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala:1840]: Confirmed in direct message

**Comment by WojciechMazur** [comment]
This PR couses the same regressions as reported in https://github.com/scala/scala3/pull/21328#issuecomment-2269094496

**Comment by EugeneFlesselle** [comment]
> This PR couses the same regressions as reported in [#21328 (comment)](https://github.com/scala/scala3/pull/21328#issuecomment-2269094496)

But only in `3.6` now right ?? the only major difference with that PR is that we delayed the changes by one version

**Comment by WojciechMazur** [comment]
It's happening in default, that is -source:3.5 compilation. I'm working on reproducers right now for missing implicits

**Comment by EugeneFlesselle** [comment]
> It's happening in default, that is -source:3.5 compilation. I'm working on reproducers right now for missing implicits

The only place emitting warnings with messages of the form _"Given search preference for ... between alternatives ... none - it's ambiguous"_ is in `compareAlternatives` under the condition that `isWarnPriorityChangeVersion` which is equal to:
```scala
val sv = Feature.sourceVersion
val isLastOldVersion = sv.stable == SourceVersion.`3.6`
val isWarnPriorityChangeVersion = isLastOldVersion || sv == SourceVersion.`3.7-migration`
```

So I am very confused about how we can still be getting these warnings where the source version is `3.5`.

**Comment by WojciechMazur** [comment]
Reproducer for `neandertech/langoustine`

```scala
object serializer:
  trait Reader[T]
  trait Writer[T]
  // Needs to be implicit val
  implicit val UnitReader: Reader[Unit] = ???
  implicit val StringReader: Reader[String] = ???
  // A way to derive instances needs to be available
  inline given superTypeReader[T: scala.reflect.ClassTag]: Reader[T] = ???
import serializer.Reader

trait Codec[T]
trait Channel[F[_]]:
  def notificationStub[In: Codec](): In => F[Unit]
trait Monadic[F[_]]

sealed abstract class LSPNotification():
  type In
  given inputReader: Reader[In]

class PreparedNotification[X <: LSPNotification](val x: X, val in: x.In):
  type In = x.In

trait Communicate[F[_]]:
  def notification[X <: LSPNotification](notif: X, in: notif.In): F[Unit]

object Communicate:
  given codec[T: Reader]: Codec[T] = ???

  def channel[F[_]: Monadic](channel: Channel[F]) =
    new Communicate[F]:
      override def notification[X <: LSPNotification](notif: X, in: notif.In): F[Unit] = 
        channel.notificationStub().apply(in) // error, does not infer notif.In type so cannot summon correct implicit
```

**Comment by WojciechMazur** [comment]
> So I am very confused about how we can still be getting these warnings where the source version is `3.5`.

The regressions are not about warnings. These are hard errors the first reproducer fails even with -source:3.4

**Comment by EugeneFlesselle** [comment]
> The regressions are not about warnings. These are hard errors the first reproducer fails even with -source:3.4

I see, thank you for the reproducer

This is caused by 8a41389dd4ea6c15f2089519ac5883c4c72a0c56, that commit is necessary only to maintain transitivity in the new comparison scheme. I will investigate what exactly is happing there. In the mean time, it should be fine to drop that commit for the purpose of the backport. We should at least relaunch the open CB on this PR -  8a41389dd4ea6c15f2089519ac5883c4c72a0c56

**Comment by EugeneFlesselle** [comment]
> Reproducer for neandertech/langoustine

Minimized a little:
```scala
trait Reader[T]

// Needs to be implicit val
implicit val UnitReader: Reader[Unit] = ???
implicit val StringReader: Reader[String] = ???
// A way to derive instances needs to be available
given superTypeReader[T: scala.reflect.ClassTag]: Reader[T] = ???

object Notif:
  type In
  given inputReader: Reader[In] = ???

object Test:

  def notificationStub[T: Reader]: T => Unit = ???

  val in: Notif.In = ???
  notificationStub.apply(in) // error since 8a41389d
```

> This is caused by [8a41389](https://github.com/scala/scala3/commit/8a41389dd4ea6c15f2089519ac5883c4c72a0c56), that commit is necessary only to maintain transitivity in the new comparison scheme. I will investigate what exactly is happing there. 

In 8a41389, we did not account for the fact that the result of `rank` also depends on the `rfailures` parameter. Specifically, once there are no more pending candidates, we do:
```scala
case nil =>
  if (rfailures.isEmpty) found
  else found.recoverWith(_ => rfailures.reverse.maxBy(_.tree.treeSize))
```

Having the split the ranking of implicits and givens changes the accumulated `rfailures` with which rank finishes, and thereby the returned `SearchFailure`. Which in turn seems to prevent from even considering the `Notif.inputReader` candidate.

**Comment by WojciechMazur** [comment]
Reproducer for `tpolcat/doobie` 

```scala
trait Text[T]
trait Read[A]
object Read extends ReadImplicits:
  implicit val unit: Read[Unit] = ???
trait ReadImplicits:
  import scala.deriving.*
  given roe: Read[Option[EmptyTuple]] = ???
  given rou: Read[Option[Unit]] = ???
  given cons1[H, T <: Tuple](using Read[Option[H]], Read[Option[T]]): Read[Option[H *: T]] = ???

trait Fragment:
  def query[B: Read]: String = ???

@main def Test = 
  val f: Fragment = ???
   f.query // error
```

This snipet was compiled previously only thanks to a big amount of luck.  Typical usage of `query` method requires to explicitly provide a type argument. Previously thanks to having `implicit val unit: Read[Unit]` as the only alternative in `object Read` this instance was chosen, which later infered whole expression to `query[Unit]`. I'm not sure if the library author expected that (probably not as I can see multiple usaged of `query[Unit]` in the codebase). 
Now it fails with: 
```scala
-- [E172] Type Error: /Users/wmazur/projects/sandbox/src/test.scala:16:10 ------
16 |   f.query // error
   |          ^
   |No best given instance of type Read[B] was found for a context parameter of method query in trait Fragment.
   |I found:
   |
   |    Read.cons1[H, T](
   |      /* ambiguous: both given instance roe in trait ReadImplicits and given instance rou in trait ReadImplicits match type Read[Option[H]] */
   |        summon[Read[Option[H]]],
   |    Read.roe)
   |
   |But both given instance roe in trait ReadImplicits and given instance rou in trait ReadImplicits match type Read[Option[H]].
1 error found
```

**Comment by WojciechMazur** [comment]
All 4 regressions are gone after reverting https://github.com/scala/scala3/commit/8a41389dd4ea6c15f2089519ac5883c4c72a0c56, it does not cause any new problems.

**Comment by odersky** [comment]
> All 4 regressions are gone after reverting https://github.com/scala/scala3/commit/8a41389dd4ea6c15f2089519ac5883c4c72a0c56, it does not cause any new problems.

OK great! I was thinking about how this could cause problems. The only thing that should be different is now we typecheck more implicits. Say we have a simple given and a complicated recursive implicit. Before https://github.com/scala/scala3/commit/8a41389dd4ea6c15f2089519ac5883c4c72a0c56, we would typecheck the given first and then the implicit. Afterwards, the the reverse. This could very well cause problems. One way to cause fewer problems is to start with the simplest term after sorting, then compare it with all terms of the same implicit-vs-given-class, then do the rest. But in general it seems impossible to guarantee no regressions when we change the order in which candidates are type checked.

**Pull request merged**